File: internal/agent/port_security.go
package agent

// OpPortSecurityPrecheck 探测一个网口的端口安全能力与将要下发的规则。**只读**。
//
// 它必须来自节点而不是控制面自己算：
//
//   - **能力是否具备**只有节点知道（OVS 装了没有、OpenFlow13 支持不支持、
//     meter 有没有）。控制面猜错的代价是给用户一个不会生效的"已启用"。
//   - 最终写下去的流表取决于该网口当前所属的网桥、已有的流表与端口号，
//     控制面按模板拼一段出来看起来对、实际可能相差很远。
const OpPortSecurityPrecheck OpKind = "port_security.precheck"

// OpPortSecurityApply 下发端口安全策略。
//
// 下发的是**期望状态**（三项保护各开不开、限速多少），而不是"增删某条规则"。
// 理由是节点上的实际状态可能已经被别人改过（手工加过流表、节点重启过），
// 增量指令在那种情况下会收敛到错误结果，而期望状态无论当前是什么样都能
// 收敛到正确结果。
const OpPortSecurityApply OpKind = "port_security.apply"

// PortSecurityPrecheckKey 是预检结果中承载详情的键。
const PortSecurityPrecheckKey = "port_security_precheck"

// PortSecurityDataKey 是下发结果中承载补充信息的键。
const PortSecurityDataKey = "port_security"

// PortSecurityPrecheck 是节点返回的预检详情。
type PortSecurityPrecheck struct {
	// Capabilities 是三项能力各自的可用状态。
	//
	// 形状与 network 包的能力清单一致（Key / Label / Required / Missing /
	// Reason / Fix），这样界面上可以直接复用同一套呈现。
	Capabilities []PortSecurityCapability
	// Rules 是将要写入的流表，人可读。
	Rules []string
	// Warnings 是节点侧发现的冲突（如该网口已有手工流表会被覆盖）。
	Warnings []string
}

// PortSecurityCapability 是一项能力的可用状态。
type PortSecurityCapability struct {
	Key      string
	Label    string
	Required bool
	Missing  bool
	Reason   string
	Fix      string
}

// PortSecurityInfo 是下发结果。
type PortSecurityInfo struct {
	// Applied 为 true 表示三项保护已按期望状态写入。
	Applied bool
	Message string
}

File: web/src/api/portsecurity.ts
/**
 * 端口安全接口（F-4-08）。
 *
 * 三项保护的性质**不同**，界面必须区别对待：
 *
 *   防伪造  安全边界，而且它是其它一切的前提——不防欺骗的话，网络里所有
 *           "按地址区分机器"的策略（隔离、防火墙、端口转发）都能被绕过。
 *           它不会让任何现有连接断开，因此**不该**被当成危险开关。
 *   隔离    也是安全边界，但代价大：同网段内**所有**机器之间都不通了，
 *           包括用户自己放在一起的应用集群。
 *   限速    资源保护，不是安全措施。默认不限——默认限流会在用户什么都没做
 *           的时候开始丢包，而那种丢包看起来像应用的问题。
 */
import { del, get, post } from './client'

export interface PortSecurityPolicyView {
  id: number
  node_id: number
  port_ref: string
  switch_id?: number
  vm_id?: number
  vm_name?: string

  spoofing_guard: boolean
  isolation: boolean
  pps_limit: number

  status: 'pending' | 'active' | 'failed'
  /** 为空表示尚未生效。与「已启用」是两回事。 */
  applied_at?: string
  detail?: string
  created_at: string
}

export interface CapabilityState {
  key: string
  label: string
  required: boolean
  missing: boolean
  reason?: string
  fix?: string
}

export interface PortSecurityPrecheck {
  /** 为 false 时**不能启用**。 */
  can_apply: boolean
  capabilities: CapabilityState[]
  /** 将要下发的规则，来自节点。 */
  rules: string[]
  warnings: string[]
}

export interface PortSecurityRequest {
  port_ref: string
  switch_id?: number
  vm_id?: number
  spoofing_guard: boolean
  isolation: boolean
  pps_limit: number
}

export const portSecurityApi = {
  list: (nodeID: number) =>
    get<{ items: PortSecurityPolicyView[] }>('/api/v1/port-security', { node_id: nodeID }),

  /** 预检：探测能力并给出将下发的规则。**只读**。 */
  preview: (nodeID: number, req: PortSecurityRequest) =>
    post<PortSecurityPrecheck>(`/api/v1/port-security/preview?node_id=${nodeID}`, req),

  /**
   * 配置并启用。`can_apply` 为 false 时服务端**直接拒绝**；
   * 有警告而未确认时**不执行也不报错**，而是返回预检结果。
   */
  apply: (nodeID: number, req: PortSecurityRequest, acknowledge: boolean) =>
    post<{ precheck: PortSecurityPrecheck; task_id?: string }>(
      `/api/v1/port-security?node_id=${nodeID}`,
      { ...req, acknowledge },
    ),

  disable: (id: number) => del<{ task_id?: string }>(`/api/v1/port-security/${id}`),
}

export const PS_STATUS_LABEL: Record<string, string> = {
  pending: '未生效',
  active: '已生效',
  failed: '失败',
}

/**
 * PS_LIMIT_BOUNDS 与服务端的边界保持一致。
 *
 * 界面自己先拦一道，是为了在**用户输入时**就给出提示，而不是等他点了
 * 提交才收到一个错误。服务端仍然会再校验一次——界面上的校验只是体验，
 * 不是约束。
 */
export const PS_LIMIT_BOUNDS = { min: 10, max: 1_000_000 }
